### PR TITLE
Chapter 5: Structs

### PR DESCRIPTION
**Pull request by caalo**
First attempt at Structs.

### New Content Checklist

- [ ] New content/chapter is in an Rmd file with [this kind of format and headers](https://github.com/jhudsl/OTTR_Template/blob/main/02-chapter_of_course.Rmd).

- [ ] New content/chapter contains learning objectives.

- [ ] [Bookdown successfully re-renders and any new content files have been added to the _bookdown.yml](https://github.com/jhudsl/OTTR_Template/wiki/Publishing-with-Bookdown).

- [ ] [Spell check runs successfully](https://www.ottrproject.org/customize-robots.html#Spell_checking)).

- [ ] Any newly necessary packages that are needed have been added to the [Dockerfile and image](https://www.ottrproject.org/customize-docker.html).

- [ ] Images are in the [correct format for rendering](https://www.ottrproject.org/writing_content.html#set-up-images).

- [ ] Every new image has [alt text and is in a Google Slide](https://www.ottrproject.org/writing_content.html#Accessibility).

- [ ] Each slide is described in the notes of the slide so learners relying on a screen reader can access the content. See https://lastcallmedia.com/blog/accessible-comics for more guidance on this.

- [ ] The color palette choices of the slide are contrasted in a way that is friendly to those with color vision deficiencies.
You can check this using [Color Oracle](https://colororacle.org/).

### CODE REVIEW

**Review by sitapriyamoorthi — APPROVED** [review]

File: 05-structs.Rmd [hunk 1]
```diff
@@ -4,11 +4,47 @@ ottrpal::set_knitr_image_path()
 
 # Organizing variables via Structs
 
+In our workflow so far, we see that certain variables are always used together, even for different tasks. For example, variables related to the reference genome are always used for the same purpose and passed on to tasks in almost the same way. This leads to quite a bit of coding redundancy, as when we write down the large set of variables related to the reference genome as task inputs, we are just thinking about one entity. We don't make distinctions of the reference genome files until the task body itself.
 
+To improve code organization and readability, we can package all variables related to the reference genome into a compound data structure called a **struct**. With a struct variable, we can refer all the packaged variables as one single variable, and also refer to specific variables within the struct without losing any information.
```
Comment (sitapriyamoorthi) [inline 05-structs.Rmd:9]: Maybe add link here to structs from Open WDL docs

File: 05-structs.Rmd [hunk 2]
```diff
@@ -4,11 +4,47 @@ ottrpal::set_knitr_image_path()
...
+  # Map reads to reference
+  call BwaMem {
+    input:
+      input_fastq = sampleFastq,
+      refGenome = refGenome
+  }
+}
 ```
+
+The `referenceGenome` struct contains all the variables related to the reference genome, but values cannot be defined. The struct definition merely lays the skeleton components of the data structure, but contains no actual values.
```
Comment (sitapriyamoorthi) [inline 05-structs.Rmd:41]: Consider changing The `referenceGenome` struct contains all the variables related to the reference genome, but values cannot be defined. 
to 
The `referenceGenome` struct contains all the variables related to the reference genome, but values cannot be defined here.

File: 05-structs.Rmd [hunk 3]
```diff
@@ -0,0 +1,467 @@
...
+    input:
+      input_fastq = sampleFastq,
+      refGenome = refGenome
+  }
+}
+```
+
+The `referenceGenome` struct contains all the variables related to the reference genome, but values cannot be defined here. The struct definition merely lays the skeleton components of the data structure, but contains no actual values.
+
+In our workflow inputs, we declare a `referenceGenome` struct variable called `refGenome` via `referenceGenome refGenome` and removed our original reference genome definitions. We can access the variables within a struct by the following syntax: `structVar.varName`, such as `refGenome.ref_name`. The [WDL syntax specs](https://github.com/openwdl/wdl/blob/main/versions/1.0/SPEC.md#struct-definition) has more information on how to define and use structs.
```
Comment (aofarrel) [inline 05-structs.Rmd:43]: I'd reword this slightly for clarity and consistent verb tense.

```
In our workflow inputs, we remove all of the `File` variables associated with reference genome definitions, but keep anything that isn't related to the reference genome, such as `sampleFastq`. We instead declare a `referenceGenome` struct variable called `refGenome` via `referenceGenome refGenome`. We can access the variables within a struct by the following syntax: `structVar.varName`, such as `refGenome.ref_name`. The [WDL spec](https://github.com/openwdl/wdl/blob/main/versions/1.0/SPEC.md#struct-definition) has more information on how to define and use structs.
```

File: 05-structs.Rmd [hunk 4]
```diff
@@ -0,0 +1,467 @@
...
+
+To give values to `refGenome`, we need to modify our JSON metadata file. We define `refGenome` variable in a nested structure that corresponds to the `referenceGenome` struct. Let's take a look:
```
Comment (aofarrel) [inline 05-structs.Rmd:45]: Consider change to "We define the `refGenome` variable..." (add "the")

File: 05-structs.Rmd [hunk 5]
```diff
@@ -0,0 +1,467 @@
...
+    "ref_name": "hg19"
+  },
+  "minidata_mutation_calling_v1.dbSNP_vcf_index": "/fh/fast/paguirigan_a/pub/ReferenceDataSets/genome_data/human/hg19/dbsnp_138.b37.vcf.gz.tbi",
+  ...
+}
+```
+
+Now `refGenome` has all the values it needs for our tasks.
+
+In addition, we have replaced all the reference genome inputs in `call BwaMem` with `refGenome`: we can pass information to a task via structs.
```
Comment (aofarrel) [inline 05-structs.Rmd:67]: Consider change to "...in `call BwaMem` with `refGenome` in order to pass information to a task via structs."

File: 05-structs.Rmd [hunk 6]
```diff
@@ -0,0 +1,467 @@
...
+
+Within the `BwaMem` task, we have to to refer to variables inside the struct, such as `refGenome.ref_name` (which has a value of "hg19" using this JSON metadata):
```
Comment (aofarrel) [inline 05-structs.Rmd:69]: Change to "Within the `BwaMem` task, we must refer to variables..." (remove "have to to")

File: 05-structs.Rmd [hunk 7]
```diff
@@ -0,0 +1,467 @@
...
+  String read_group_id = "ID:" + base_file_name
+  String sample_name = "SM:" + base_file_name
+  String platform = "illumina"
+  String platform_info = "PL:" + platform   # Create the platform information
+
+
+  command <<<
+    set -eo pipefail
+
+    #can we iterate through a struct??
```
Comment (aofarrel) [inline 05-structs.Rmd:90]: Yes, but unlike an array it has to be done by key, so it's a little goofy. This should do the trick:

```
ref_files=("~{refGenome.ref_fasta}" "~{refGenome.ref_fasta_index}" "~{refGenome.ref_dict}" "~{refGenome.ref_amb}" "~{refGenome.ref_ann}" "~{refGenome.ref_bwt}" "~{refGenome.ref_pac}" "~{refGenome.ref_sa}")
for ref_file in "${ref_files[@]}"
do 
    mv "$ref_file" .
done
```

File: 05-structs.Rmd [hunk 8]
```diff
@@ -0,0 +1,467 @@
...
+  
+  runtime {
+    memory: "48 GB"
+    cpu: 16
+    docker: "fredhutch/bwa:0.7.17"
+  }
+}
+```
+
+Other tasks in the workflow, such as `ApplyBaseRecalibrator` and `Mutect2TumorOnly` also make use of the reference genome, we so pass `refGenome` to it. For the final task `annovar`, it only requires the reference genome name, and none of the files in the `referenceGenome` struct. We make a stylistic choice to pass only `refGenome.ref_name` to the input of `annovar` task calling, as the task doesn't need the full information of the struct. This stylistic choice is based on the principle of passing on the minimally needed information for a modular piece of code to run, which makes the task `annovar` depend on the minimal amount of inputs.
```
Comment (aofarrel) [inline 05-structs.Rmd:119]: Changes to consider:
* "...genome, so we pass `refGenome` to it..."
* "The final task `annovar` only requires..."
* "...to the input of `annovar` task call, as the task..." (whether this noun should be "call" or "calling" is probably a matter of opinion, but I usually see call as the noun and verb)

Also consider adding "This will also save us time and disk space by not having to localize several gigabytes of reference files into the Docker container that `annovar` will be running in." at the very end.

File: 05-structs.Rmd [hunk 9]
```diff
@@ -0,0 +1,467 @@
...
+  String read_group_id = "ID:" + base_file_name
+  String sample_name = "SM:" + base_file_name
+  String platform = "illumina"
+  String platform_info = "PL:" + platform   # Create the platform information
+
+
+  command <<<
+    set -eo pipefail
+
+    #can we iterate through a struct??
```
Comment (aofarrel) [inline 05-structs.Rmd:247]: If we change the struct iteration above, change it here too.

File: 05-structs.Rmd [hunk 10]
```diff
@@ -0,0 +1,467 @@
...
+  
+  String ref_fasta_local = basename(refGenome.ref_fasta)
+  String dbSNP_vcf_local = basename(dbSNP_vcf)
+  String known_indels_sites_VCFs_local = basename(known_indels_sites_VCFs)
+
+
+  command <<<
+  set -eo pipefail
+
+  mv ~{refGenome.ref_fasta} .
```
Comment (aofarrel) [inline 05-structs.Rmd:331]: we should be able to make a bash array with all of these files, including dbSNP_vcf, known_indels_sites_VCFs, etc.

File: 05-structs.Rmd [hunk 11]
```diff
@@ -0,0 +1,467 @@
...
+  }
+
+    String base_file_name = basename(input_bam, ".recal.bam")
+    String ref_fasta_local = basename(refGenome.ref_fasta)
+    String genomeReference_local = basename(genomeReference)
+
+command <<<
+    set -eo pipefail
+
+    mv ~{refGenome.ref_fasta} .
```
Comment (aofarrel) [inline 05-structs.Rmd:396]: same principle here r/e bash arrays

File: 05-structs.Rmd (hunk 7 above)
Comment (caalo) [reply 05-structs.Rmd:90]: Awesome thanks. This is something bigger that we should decide whether to use it for all of our file localization commands. Going to put this in an issue and we can talk more about it!
